Create Dark Mode Modal - RSC-1171

EmilyBermudez88 opened the pull request:
https://issues.sonatype.org/browse/RSC-1171

I currently have two different methods of how to add the various classes to the `html` element.
Although the first option using `classnames()` seems a bit cleaner, I have not yet figured out a good way to prevent the initial classes the `html` element accepts (`nx-html` and `nx-html--page-scrolling`) from being overridden without hardcoding them into classnames. I'm not sure if this is the best idea to override what is currently laid out in the index.html file (even if the classes are the same).
The second / commented-out option has more lines of code, but is simpler by using `classList.add / remove` in a variety of scenarios. 
I can continue fine tuning either option if needed, but wanted an opinion of which approach I should focus first.

rpokorny commented:
Something more along the lines of the second option I think is what we want - it's similar to how we've implemented `SectionScrollingWrapper` (which you'll want to make sure you don't break!).  However I think you could make the code more succinct, especially by using the two-parameter version of the [`classList.toggle`](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle) function

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,133 @@
+  }, [enableModeChange, mode]);
+
+  useEffect(()=> {
+    enableModeChange ? setRadioDisabled(false) : setRadioDisabled(true);
```
Whether the radiogroup is disabled is strictly based on `enableModeChange`, so you don't want `radioDisabled` to be its own piece of state at all – it can just be a local variable

EmilyBermudez88 replied:
> Whether the radiogroup is disabled is strictly based on `enableModeChange`, so you don't want `radioDisabled` to be its own piece of state at all – it can just be a local variable

I updated it here : [a91181a](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/a91181a9d5387764cb85f29a9d5bd4cb4b563dde). Since disabled is always the opposite of enableModeChange, I assigned the disabled attribute the opposite value of enableModeChange

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,133 @@
+
+  return (
+    <>
+      <NxButton variant="tertiary" onClick={() => setShowModal(true)}>Light / Dark Mode</NxButton>
```
How about something like "Theme Settings" for the button label?

EmilyBermudez88 replied:
> How about something like "Theme Settings" for the button label?

Sounds good, replaced here: [2dd372b](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/2dd372bd8036da861ef9ef7d380a830a86e7abf2)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,133 @@
+      <NxButton variant="tertiary" onClick={() => setShowModal(true)}>Light / Dark Mode</NxButton>
+      { showModal &&
+        <NxModal id="nx-modal-dark-mode-example"
+                 role="alertdialog"
```
The default `dialog` role seems more correct to this use case

EmilyBermudez88 replied:
> The default `dialog` role seems more correct to this use case

changed to `dialog` here: [92a7597](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/92a75978913d80a5b8101f5aef71e4c7307c81e7)

rpokorny replied:
> changed to dialog here: [92a7597](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/92a75978913d80a5b8101f5aef71e4c7307c81e7)

My intent was that you remove `role` here entirely, as `dialog` is already the default

EmilyBermudez88 replied:
> My intent was that you remove `role` here entirely, as `dialog` is already the default

Removed here: [3cfdc2c](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/3cfdc2c2d1a89c6225304e8474917681acdd9c19)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,133 @@
+        <NxModal id="nx-modal-dark-mode-example"
+                 role="alertdialog"
+                 onCancel={modalCloseHandler}
+                 aria-label="NxModal to set light or dark mode">
```
Use `aria-labelledby` to reference the `h2` instead of `aria-label`

EmilyBermudez88 replied:
> Use `aria-labelledby` to reference the `h2` instead of `aria-label`

changed to `aria-lablledby` here: [92a7597](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/92a75978913d80a5b8101f5aef71e4c7307c81e7)

rpokorny replied:
> changed to aria-lablledby here: 92a7597

The id in question should really be more specific to this modal too

EmilyBermudez88 replied:
> The id in question should really be more specific to this modal too

Renamed here: [227bab8](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/227bab8fc6853c6e335f2877b23e1b86b482bee5)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,133 @@
+                  submitBtnText= "Save and Close">
+            <header className="nx-modal-header">
+              <h2 className="nx-h2" id="modal-form-header">
+                <span>Setting Preferences for Light or Dark Mode</span>
```
```suggestion
                Setting Preferences for Light or Dark Mode
```

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,133 @@
+                simultaneously will result in dark mode and is not recommended. If neither class is added,
+                prefers-color-scheme in the user’s settings will dictate the mode.
+              </NxP>
+              <NxFormGroup label="Enable Changes">
+                <NxToggle onChange={toggleEnableModeChange} isChecked={enableModeChange}>
+                  Opt-in to Allow Changes
```
"Changes" seems a little bit meaningless here, how about "Theming"

EmilyBermudez88 replied:
> "Changes" seems a little bit meaningless here, how about "Theming"

Updated here: [ba3a3df](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/ba3a3df0df0eecdf0fdaeb0141e6f465f8a44415)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,133 @@
+            </header>
+            <div className="nx-modal-content">
+              <NxP>
+                In order to enable dark mode, the class <NxCode>nx-html--enable-color-schemes</NxCode> must first
+                be added to the <NxCode>html</NxCode> element, as demonstrated by the toggle below. Additional
+                classes <NxCode>nx-html--dark-mode</NxCode> or <NxCode>nx-html--light-mode</NxCode> may also be added
+                to force either light or dark mode, as demonstrated by the radio buttons. However, using these
+                simultaneously will result in dark mode and is not recommended. If neither class is added,
+                prefers-color-scheme in the user’s settings will dictate the mode.
```
I'd like the explanation here to be understandable both to technical and non-technical people.  To do that, I think it should focus primarily on the conceptual mechanisms at play (e.g. that theming is opt-in and you always get light mode if you do not opt in, and also that when opted in your browser and/or OS theme choices determine the theme by default) but with brief mention of the technical implementations (classnames and media queries) as well.

EmilyBermudez88 replied:
> I'd like the explanation here to be understandable both to technical and non-technical people. To do that, I think it should focus primarily on the conceptual mechanisms at play (e.g. that theming is opt-in and you always get light mode if you do not opt in, and also that when opted in your browser and/or OS theme choices determine the theme by default) but with brief mention of the technical implementations (classnames and media queries) as well.

As discussed in WW, the paragraph has been removed, and explanations implemented in tooltips here: [4e6b7de](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/4e6b7deaa4d5e26ea2c2c25a0722bbe7ee68cbc8)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,133 @@
+                to force either light or dark mode, as demonstrated by the radio buttons. However, using these
+                simultaneously will result in dark mode and is not recommended. If neither class is added,
+                prefers-color-scheme in the user’s settings will dictate the mode.
+              </NxP>
```
One idea I had about presenting the descriptions here was to use "information" icons that produce tooltips on hover for each of the different options, rather than explaining everything in a single paragraph at the top.

EmilyBermudez88 replied:
> One idea I had about presenting the descriptions here was to use "information" icons that produce tooltips on hover for each of the different options, rather than explaining everything in a single paragraph at the top.

Tooltips implemented here: [4e6b7de](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/4e6b7deaa4d5e26ea2c2c25a0722bbe7ee68cbc8)

rpokorny commented on gallery/src/PageHeader/PageHeader.tsx:
```diff
@@ -29,7 +30,9 @@ const PageHeader = () => (
         React Shared Component Library
       </span>
-
+      <div className="gallery-page-header__display-mode-btn">
```
Is the extra div needed?

EmilyBermudez88 replied:
> Is the extra div needed?

Nope, it works fine without. Removed here: [da74491](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/da7449165643e62842d939367922afd0b4c11f25)

EmilyBermudez88 commented:
> Something more along the lines of the second option I think is what we want - it's similar to how we've implemented SectionScrollingWrapper (which you'll want to make sure you don't break!). However I think you could make the code more succinct, especially by using the two-parameter version of the [classList.toggle](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle) function

Thanks for pointing me towards the two-parameter version of classList.toggle, I didn't know this second parameter existed, and has really cleaned things up! I removed the useEffect hooks surrounding the toggles because the dependencies were identical to the conditions of the second parameter in classList.toggle. However, I can re-implement them if you think it's better to leave them. Code updated here: [d42799f](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/d42799fb0d782a63c6d14c3a93afd4d92e768362)

rpokorny commented:
> I removed the useEffect hooks surrounding the toggles because the dependencies were identical to the conditions of the second parameter in classList.toggle. However, I can re-implement them if you think it's better to leave them. Code updated here: [d42799f](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/d42799fb0d782a63c6d14c3a93afd4d92e768362)

It still needs to be in a `useEffect`.  All side-effects should be in `useEffect` as a general rule

EmilyBermudez88 commented:
>> I removed the useEffect hooks surrounding the toggles because the dependencies were identical to the conditions of the second parameter in classList.toggle. However, I can re-implement them if you think it's better to leave them. Code updated here: [d42799f](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/d42799fb0d782a63c6d14c3a93afd4d92e768362)

>It still needs to be in a useEffect. All side-effects should be in useEffect as a general rule

Re-implemented here: [afa9fb9](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/afa9fb9ab6107c532fcb23f7ffc5f98cb74e32b7)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,111 @@
+
+  useEffect(() => {
+    classList.toggle('nx-html--enable-color-schemes', enableModeChange);
+  }, [enableModeChange]);
+
+  useEffect(() => {
```
I'd just combine these
```suggestion
```

EmilyBermudez88 replied:
> I'd just combine these

moved into one here: [803b3a0](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/803b3a0a0d10d335b14f55ba3ce34acb4c51a022)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,111 @@
+  const [enableModeChange, toggleEnableModeChange] = useToggle(false);
+  const [mode, setMode] = useState<string | null>(null);
+
+  const { classList } = document.documentElement;
```
It would be cleaner with this line moved into the `useEffect` as well

EmilyBermudez88 replied:
> It would be cleaner with this line moved into the `useEffect` as well

moved into the useEffect here: [803b3a0](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/803b3a0a0d10d335b14f55ba3ce34acb4c51a022)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,111 @@
+                  Opt-in to Allow Theming
+                  <NxTooltip title="Opting in will allow you to make adjustments to the display theme. Failure
+                             to do so will always result in light mode. This can be done by adding the
+                             'nx-html--enable-color-schemes' class to the HTML element.">
```
I think we could phrase these in a way that makes it more clear that toggling the input will in fact make the change to the classname; the current phrasing almost makes it sound like the user could/should do something _in addition_ to toggling the toggle.  For instance, this one could say "This operates by toggling the presence of the nx-html--enable-color-schemes class"

EmilyBermudez88 replied:
> I think we could phrase these in a way that makes it more clear that toggling the input will in fact make the change to the classname; the current phrasing almost makes it sound like the user could/should do something _in addition_ to toggling the toggle. For instance, this one could say "This operates by toggling the presence of the nx-html--enable-color-schemes class"

Thanks for pointing that out, I hadn't thought about it that way. It has been changed here: [e43333b](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/e43333b0e5ea78c9c661c52d2485325d30fdd5c9)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,111 @@
+                  Opt-in to Allow Theming
+                  <NxTooltip title="Opting in will allow you to make adjustments to the display theme. Failure
+                             to do so will always result in light mode. This can be done by adding the
+                             'nx-html--enable-color-schemes' class to the HTML element.">
```
Note that tooltip titles can be JSX.  I think it would be worth trying `NxCode` where appropriate within the tooltip to see how it looks.

EmilyBermudez88 replied:
> Note that tooltip titles can be JSX. I think it would be worth trying `NxCode` where appropriate within the tooltip to see how it looks.

 I did try it initially, but wasn't about it's appearance. Here's a screenshot of what it looks like inside the tooltip:
![Screen Shot 2022-09-22 at 8 49 54 AM](https://user-images.githubusercontent.com/89862727/191754650-f8c0bbaf-50d4-4fe5-a9d3-415807cf5c6f.png)

rpokorny replied:
> I did try it initially, but wasn't sure about if it looked ok. Here's a screenshot of what it looks like inside the tooltip:

Check with design

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,111 @@
+                         isChecked={mode === 'dark'}
+                         disabled = {!enableModeChange}>
+                  Dark Mode
+                  <NxTooltip title="Override the browser's default display theme to dark mode by adding the class
```
```suggestion
                  <NxTooltip title="Overrides the browser's default display theme to dark mode by adding the class
```

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,111 @@
+                         isChecked={mode === 'light'}
+                         disabled = {!enableModeChange}>
+                  Light Mode
+                  <NxTooltip title="Override the browser's default display theme to light mode by adding the class
```
```suggestion
                  <NxTooltip title="Overrides the browser's default display theme to light mode by adding the class
```

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,111 @@
+                         isChecked={mode === 'browserChoice'}
+                         disabled = {!enableModeChange}>
+                  Let Your Browser Color Preference Decide
+                  <NxTooltip title="If theme changes are permitted, the default theme will be dictated by
+                             your browser or OS theme choice. The prefers-color-scheme media query registers
+                             this choice.">
```
I think we can remove the conditional aspect of this sentence.  The other two radios aren't described conditionally, and IMO the fact that it's conditional is made clear by the radios becoming disabled when the Opt-In toggle is off.

EmilyBermudez88 replied:
> I think we can remove the conditional aspect of this sentence. The other two radios aren't described conditionally, and IMO the fact that it's conditional is made clear by the radios becoming disabled when the Opt-In toggle is off.

Absolutely, changed here: [3f5e143](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/3f5e1437e88f0edbdce9ae5bd7708d4676a6d0f4)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,111 @@
+                         disabled = {!enableModeChange}>
+                  Let Your Browser Color Preference Decide
+                  <NxTooltip title="If theme changes are permitted, the default theme will be dictated by
+                             your browser or OS theme choice. The prefers-color-scheme media query registers
+                             this choice.">
```
```suggestion
                             your browser or OS theme choice, which is communicated to the application via the 'prefers-color-scheme' media query.">
```

EmilyBermudez88 replied:
Browser choice text changed here: [3f5e143](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/3f5e1437e88f0edbdce9ae5bd7708d4676a6d0f4)

rpokorny commented on gallery/src/PageHeader/PageHeader.scss:
```diff
@@ -37,7 +37,7 @@
   }
 
-  &__version {
+  .nx-btn--tertiary {
```
In this case I think it would be better to select the button either by an id or a very specific classname: we wouldn't want this style to apply to any other hypothetical tertiary button in the page header.

EmilyBermudez88 replied:
> In this case I think it would be better to select the button either by an id or a very specific classname: we wouldn't want this style to apply to any other hypothetical tertiary button in the page header.

I added a classname here: [357cbcd](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/357cbcda72bcf2e87d6c013223fa333f5a32ee8e)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,111 @@
+
+  return (
+    <>
+      <NxButton variant="tertiary" onClick={() => setShowModal(true)}>Theme Settings</NxButton>
```
It's a bit atypical to have the button that opens the modal be in the same component as the modal.  Let's move the button itself and the state it controls to the `PageHeader` component

EmilyBermudez88 replied:
> It's a bit atypical to have the button that opens the modal be in the same component as the modal. Let's move the button itself and the state it controls to the `PageHeader` component

Sorry, this one came with a lot of changes. With the button moving to `PageHeader`, the settings inside the modal reset each time it was opened. So there is now an additional useEffect to check if various classnames are present, and set the mode accordingly. I removed state that controlled the toggling of `enableModeChange`, in favour of one `mode` state that has a `disabled` option as well.  [c5cda2c](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/c5cda2ca1dc81c9a3682c94dc0073bf3c1f749e0)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,111 @@
+                         onChange={setMode}
+                         isChecked={mode === 'browserChoice'}
+                         disabled = {!enableModeChange}>
+                  Let Your Browser Color Preference Decide
```
Wrap the text for each of these in a span, then you won't need the CSS customization.  This is why we have a general rule not to mix bare text nodes and elements as siblings unless the elements are simple inline elements.

EmilyBermudez88 replied:
> Wrap the text for each of these in a span, then you won't need the CSS customization. This is why we have a general rule not to mix bare text nodes and elements as siblings unless the elements are simple inline elements.

Spans added here: [033ffd8](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/033ffd88cf7728ebcbd52bb1eb3fddef3e7b99e7)

rpokorny commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,111 @@
+        <NxModal id="nx-modal-dark-mode-example"
+                 onCancel={modalCloseHandler}
+                 aria-labelledby="nx-modal-dark-mode-form-header">
+          <NxForm onSubmit={modalCloseHandler}
```
The behavior and presentation around when the settings take effect is inconsistent.  The fact that the footer button is a primary button, tied to a form submission, with text "Save and Close" would seem to imply that the settings in this modal do not take effect until you click that button and close the modal.  On the other hand, the use of `NxToggle` implies an immediate change, and in fact you have the logic set up to make the changes immediately.  So we need to rectify that difference, either by making the changes not take effect until the form is submitted (and also by using `NxCheckbox` instead of `NxToggle`), or we need to make it more clear that the settings don't wait for any sort of submission, probably by removing the `NxForm` wrapper entirely and instead creating a simple `NxFooter` with a secondary or tertiary button that's just labelled "Close".

EmilyBermudez88 replied:
> The behavior and presentation around when the settings take effect is inconsistent. The fact that the footer button is a primary button, tied to a form submission, with text "Save and Close" would seem to imply that the settings in this modal do not take effect until you click that button and close the modal. On the other hand, the use of `NxToggle` implies an immediate change, and in fact you have the logic set up to make the changes immediately. So we need to rectify that difference, either by making the changes not take effect until the form is submitted (and also by using `NxCheckbox` instead of `NxToggle`), or we need to make it more clear that the settings don't wait for any sort of submission, probably by removing the `NxForm` wrapper entirely and instead creating a simple `NxFooter` with a secondary or tertiary button that's just labelled "Close".

Opted to get rid of the form, which made more sense IMO - [e7e6fad](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/e7e6fad4ba850da86da7c7a127234dab4dced720)

rpokorny commented:
It'd be really great if this setting was remembered across refreshes.  Look into `localstorage` as a way to accomplish that.

EmilyBermudez88 commented:
> It'd be really great if this setting was remembered across refreshes. Look into `localstorage` as a way to accomplish that.

So I managed to get `localstorage` up and running, which I used to store the classnames to apply to the `html` element. I currently have the stored values being pulled down in Application.tsx so that the settings are applied on each page load. If there's a better place for that, I can move 'localStorage.getItem` elsewhere.
Implemented here:  [f59f81d](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/f59f81d3810be1825b254762db7ffa63077e4e44) and [3f9848d](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/3f9848dbbd57a40cde384ed08e5aa662b3b685e2)

rpokorny commented:
> So I managed to get localstorage up and running, which I used to store the classnames to apply to the html element. I currently have the stored values being pulled down in Application.tsx so that the settings are applied on each page load. If there's a better place for that, I can move localStorage.getItem elsewhere.
Implemented here: [f59f81d](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/f59f81d3810be1825b254762db7ffa63077e4e44) and [3f9848d](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/3f9848dbbd57a40cde384ed08e5aa662b3b685e2)

I got halfway through some suggested changes to move the localstorage management into its own file, but unfortunately I really have to leave now and will be out next week.  Therefore I've put those changes on this branch: https://github.com/sonatype/sonatype-react-shared-components/compare/RSC-1171_create_dark_mode_toggles...RSC-1171_create_dark_mode_toggles_ross. I suspect you'll be able to finish them up.

entroform commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,116 @@
+
+  useEffect(function() {
+    setThemingEnabled(themingEnabledState);
+  }, [themingEnabled, themingEnabledState]);
```
```suggestion
  }, [themingEnabledState]);
```

EmilyBermudez88 commented:
I noticed that the `nx-html--dark-mode` and `nx-html--light-mode` classes were not being removed when themingEnabled was set to false (aka toggle set to false), so I added in condition for when the classes are applied here : 
[f1e5038](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/f1e5038f715a1204b42e6c0c67055701244249d1)

entroform commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,116 @@
+  modalCloseHandler: () => void;
+};
+
+const DarkModeModal = (props:Props) => {
+  const { modalCloseHandler } = props;
```
```suggestion
const DarkModeModal = ({ modalCloseHandler }: Props) => {
```

entroform commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,116 @@
+  modalCloseHandler: () => void;
+};
+
+const DarkModeModal = (props:Props) => {
+  const { modalCloseHandler } = props;
```
Actually, I wonder if it's a good idea to rename this prop to `onClose` so it matches with the other callback names.

EmilyBermudez88 replied:
> Actually, I wonder if it's a good idea to rename this prop to `onClose` or `onCancel` so it matches with the other callback names.

I updated the naming here: [0877b05](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/0877b05ef19497dd25364eb2786a988810bf501b). I did keep the name 'onModalCloseHandler' inside pageHeader, since I think it helps distinguish what it does in the context of the page header. If you'd like me to change that one too, I can.

entroform commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,116 @@
+      </header>
+      <div className="nx-modal-content">
+        <NxFormGroup label="Enable Theme Changes">
+          <NxToggle onChange={()=> setThemingEnabledState(!themingEnabledState)} isChecked={themingEnabledState}>
```
```suggestion
          <NxToggle onChange={() => setThemingEnabledState(!themingEnabledState)} isChecked={themingEnabledState}>
```

entroform commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,116 @@
+const DarkModeModal = (props:Props) => {
+  const { modalCloseHandler } = props;
+
+  const [themingEnabledState, setThemingEnabledState] = useState(themingEnabled),
```
I wonder if we could use `useToggle` on the themingEnabledState?

EmilyBermudez88 replied:
> I wonder if we could use useToggle on the themingEnabledState?

changed to useToggle here: [9e94c49](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/9e94c49f66560c7ce74771f733b5496b3582a7db)

entroform commented on gallery/src/PageHeader/PageHeader.tsx:
```diff
@@ -4,37 +4,52 @@
-        <HostedVersionsSelect />
+const PageHeader = () => {
+  const [showModal, setShowModal] = useState(false);
+  const modalCloseHandler = () => setShowModal(false);
```
I wonder if we could use `useToggle` here too for `setShowModal`, then we can pass in the `toggleDarkModeModal` fn directly into DarkModeModal modalCloseHandler prop below

EmilyBermudez88 replied:
> I wonder if we could use `useToggle` here too for `setShowModal`, then we can pass in the `toggleDarkModeModal` fn directly into DarkModeModal modalCloseHandler prop below

Makes sense and cleaned things up nicely. Updated here: [861f2c6](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/861f2c675042b530e22789e628967d0fbaa446c8)

entroform commented on gallery/src/PageHeader/PageHeader.tsx:
```diff
@@ -4,37 +4,52 @@
+        <span className="gallery-page-header__name">
+          React Shared Component Library
+        </span>
+        <NxButton className="nx-btn--open-theme-settings"
```
Instead of a modifier `nx-btn--open-theme-settings` which implies changes in behaviour or state or is a variant of, and since this style is being used to adjust spacing within the page header, I wonder if we should change the class name to make it part of the page header:
Maybe something like `gallery-page-header__theme-settings-button` 🤔

EmilyBermudez88 replied:
> Instead of a modifier `nx-btn--open-theme-settings` which implies changes in behaviour or state or is a variant of, and since this style is being used to adjust spacing within the page header, I wonder if we should change the class name to make it part of the page header: Maybe something like `gallery-page-header__theme-settings-button` 🤔

That's a great point and something I hadn't thought of. Renamed here: [6118e55](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/6118e559989abac6643ed4a9d36d2e06069061bb)

entroform commented on gallery/src/themeClassManager.ts:
```diff
@@ -0,0 +1,47 @@
+export let themingEnabled: boolean = localStorage.getItem('theming-enabled') === 'true',
+    themeOverride: ThemeOverride = getThemeOverride();
+
+function getThemeOverride(): ThemeOverride {
```
This is a super small nitpick, I wonder if we can reorder these functions so they don't seem arbitrarily ordered: Maybe something like 🤔 :

- getThemeOverride
- setThemingEnabled
- setThemingOverride
- updateHtmlClasses

or

- setThemingEnabled
- setThemingOverride
- getThemeOverride
- updateHtmlClasses

Let me know what you think!

entroform commented:
I wonder if we should rename `DarkModeModal` to `ThemeSettingsModal` or something like that?

entroform commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,115 @@
+                to do so will always result in light mode. This operates by toggling the presence of
+                the <NxCode>nx-html--enable-color-schemes</NxCode> class.
+              </>}>
+              <NxFontAwesomeIcon className="dark-mode-modal__info-icon" icon={faInfoCircle}/>
```
I wonder if this classname `dark-mode-modal__info-icon` is necessary here and below

EmilyBermudez88 replied:
> I wonder if this classname `dark-mode-modal__info-icon` is necessary here and below

Thanks for catching that. I thought I needed it when initially building, and forgot to remove. Removed here: [4780ec8](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/4780ec883c67eeff90b0d1148ab0ff658756fb01)

entroform commented on gallery/src/DarkModeModal/DarkModeModal.tsx:
```diff
@@ -0,0 +1,116 @@
+};
+
+const DarkModeModal = ({ onClose }: Props) => {
+
```
```suggestion
```

EmilyBermudez88 commented:
> I wonder if we should rename DarkModeModal to ThemeSettingsModal or something like that, let me know what you think!

Im good with that! Renamed here: 
[afa1076](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/afa1076e5726f9e50cbe2895df313e49deb28ab9)

entroform commented on gallery/src/ThemeSettingsModal/ThemeSettingsModal.tsx:
```diff
@@ -0,0 +1,115 @@
+
+import { themingEnabled, themeOverride, setThemingEnabled, setThemeOverride } from '../themeClassManager';
+
+type Props = {
```
Small thing but we could also move this into its own `types.ts` to follow our pattern

entroform replied:
Ops nvm these aren't entirely consistent in the gallery.

entroform commented on gallery/src/ThemeSettingsModal/ThemeSettingsModal.tsx:
```diff
@@ -0,0 +1,115 @@
+  }, [themeOverrideState]);
+
+  return (
+    <NxModal id="nx-modal-dark-mode-example"
```
I'm not sure if this id is needed.

EmilyBermudez88 replied:
> I'm not sure if this id is needed.

My thought was to have an id to make it easier for testing as an easy selector. However, I can remove it if you prefer!

entroform replied:
> > I'm not sure if this id is needed.
> 
> My thought was to have an id to make it easier for testing as an easy and specific selector. However, I can remove it if you prefer!

Oo hmm, I wonder if we could remove this id and add a classname instead, maybe something like `gallery-theme-settings-modal`

EmilyBermudez88 replied:
> > > I'm not sure if this id is needed.
> > 
> > 
> > My thought was to have an id to make it easier for testing as an easy and specific selector. However, I can remove it if you prefer!
> 
> Oo hmm, I wonder if we could remove this id and add a classname instead, maybe something like `gallery-theme-settings-modal`

Changed here: [af68c48](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/af68c4892214e891dcd143b2b0fa1fc4c71c04e2)

entroform commented on gallery/src/ThemeSettingsModal/ThemeSettingsModal.tsx:
```diff
@@ -0,0 +1,115 @@
+             onCancel={onClose}
+             aria-labelledby="nx-modal-dark-mode-header">
+      <header className="nx-modal-header">
+        <h2 className="nx-h2" id="nx-modal-dark-mode-header">
```
for consistency I wonder if we should also rename this to something like `gallery-theme-settings-modal-header`

EmilyBermudez88 replied:
> for consistency I wonder if we should also rename this to something like `gallery-theme-settings-modal-header`

Thanks for clarifying about the prefix. Changed here: [af68c48](https://github.com/sonatype/sonatype-react-shared-components/pull/782/commits/af68c4892214e891dcd143b2b0fa1fc4c71c04e2)

entroform commented on gallery/src/themeClassManager.ts:
```diff
@@ -0,0 +1,46 @@
+  classList.toggle('nx-html--enable-color-schemes', themingEnabled);
+  classList.toggle('nx-html--dark-mode', themeOverride === 'dark' && themingEnabled);
+  classList.toggle('nx-html--light-mode', themeOverride === 'light' && themingEnabled);
+}
```
```suggestion
}

```

entroform reviewed (approved):
Thanks! LGTM 🚀

entroform commented on gallery/src/PageHeader/PageHeader.scss:
```diff
@@ -37,7 +37,7 @@
   }
 
-  &__version {
+  .gallery-page-header__theme-settings-button {
```
```suggestion
  &__theme-settings-button {
```
